dask/array/rechunk.py

```diff
@@ -205,6 +209,264 @@ def rechunk(x, chunks):
+    )
+
+    largest_block_size = reduce(mul, old_largest_width)
+
+    chunks = list(old_chunks)
+    memory_limit_hit = False
+
+    for dim in sorted_candidates:
+        # Examine this dimension for possible graph reduction
+        new_largest_block_size = (
+            largest_block_size * new_largest_width[dim] // old_largest_width[dim])
+        if new_largest_block_size <= block_size_limit:
+            # Full replacement by new chunks is possible
+            chunks[dim] = new_chunks[dim]
+            largest_block_size = new_largest_block_size
+        else:
+            # Try a partial rechunk, dividing the new chunks into
+            # smaller pieces
+            largest_width = old_largest_width[dim]
+            chunk_limit = int(block_size_limit * largest_width / largest_block_size)
+            c = divide_to_width(new_chunks[dim], chunk_limit)
+            if len(c) <= len(old_chunks[dim]):
+                # We manage to reduce the number of blocks, so do it
+                chunks[dim] = c
+                largest_block_size = largest_block_size * max(c) // largest_width
+
+            memory_limit_hit = True
+
+    assert largest_block_size == _largest_block_size(chunks)
+    assert largest_block_size <= block_size_limit
+    return tuple(chunks), memory_limit_hit
+
+
+def find_split_rechunk(old_chunks, new_chunks, graph_size_limit):
+    """
+    Find an intermediate rechunk that would split some chunks to
+    get us nearer *new_chunks*, without violating the *graph_size_limit*.
+    """
+    ndim = len(old_chunks)
+
+    chunks = list(old_chunks)
+
+    for dim in range(ndim):
+        graph_size = estimate_graph_size(chunks, new_chunks)
+        if graph_size > graph_size_limit:
+            break
+        if len(old_chunks[dim]) > len(new_chunks[dim]):
+            # It's not interesting to split
+            continue
+        # Merge the new chunks so as to stay within the graph size budget
+        max_number = int(len(old_chunks[dim]) * graph_size_limit / graph_size)
+        c = merge_to_number(new_chunks[dim], max_number)
+        assert len(c) <= max_number
+        # Consider the merge successful if its result has a greater length
+        # and smaller max width than the old chunks
+        if len(c) >= len(old_chunks[dim]) and max(c) <= max(old_chunks[dim]):
+            chunks[dim] = c
+
+    return tuple(chunks)
+
+
+def plan_rechunk(old_chunks, new_chunks, itemsize,
+                 threshold=4, block_size_limit=1e8):
+    """ Plan an iterative rechunking from *old_chunks* to *new_chunks*.
+    The plan aims to minimize the rechunk graph size.
+
+    Parameters
+    ----------
+    itemsize: the item size of the array
+    threshold: the graph growth factor under which we don't bother
+        introducing an intermediate step
+    block_size_limit: the maximum block size (in bytes)
+        we want to produce during an intermediate step
```

Comment:
We use the `numpydoc` standard for docstrings. The format is described [here](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections) (point 4 for parameters). I'd reformat this as:

```python
"""
...
itemsize : int
    The item size of the array
...
```
Formatting notes:
- Space between parameter and `:`
- `type` on parameter line
- Description starts with a capital letter, on line below parameter.

I also think an example section might be nice here to clarify use, and show what a returned plan might look like.